rust/transaction-filter/src/filter_operator.rs

```diff
@@ -0,0 +1,256 @@
+use crate::{
+    errors::FilterError,
+    filters::{EventFilter, TransactionRootFilter, UserTransactionFilter},
+    traits::Filterable,
+};
+use aptos_protos::transaction::v1::{transaction::TxnData, Transaction};
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+
+/// These are filters we would expect to be exposed via API
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+#[serde(tag = "type")]
+pub enum APIFilter {
+    TransactionRootFilter(TransactionRootFilter),
+    UserTransactionFilter(UserTransactionFilter),
+    EventFilter(EventFilter),
+}
+
+impl Filterable<Transaction> for APIFilter {
+    fn validate_state(&self) -> Result<(), FilterError> {
+        match self {
+            APIFilter::TransactionRootFilter(filter) => filter.is_valid(),
+            APIFilter::UserTransactionFilter(filter) => filter.is_valid(),
+            APIFilter::EventFilter(filter) => filter.is_valid(),
+        }
+    }
+
+    fn is_allowed(&self, txn: &Transaction) -> bool {
+        match self {
+            APIFilter::TransactionRootFilter(filter) => filter.is_allowed(txn),
+            APIFilter::UserTransactionFilter(ut_filter) => ut_filter.is_allowed(txn),
+            APIFilter::EventFilter(events_filter) => {
+                if let Some(txn_data) = &txn.txn_data {
+                    let events = match txn_data {
+                        TxnData::BlockMetadata(bm) => &bm.events,
+                        TxnData::Genesis(g) => &g.events,
+                        TxnData::StateCheckpoint(_) => return false,
+                        TxnData::User(u) => &u.events,
+                        TxnData::Validator(_) => return false,
+                    };
+                    events_filter.is_allowed_vec(events)
+                } else {
+                    false
+                }
+            },
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum FilterOperator {
+    And(LogicalAnd),
+    Or(LogicalOr),
+    Not(LogicalNot),
+    Filter(APIFilter),
+}
+
+impl FilterOperator {
+    pub fn new_and(and: Vec<FilterOperator>) -> Self {
+        FilterOperator::And(LogicalAnd { and })
+    }
+
+    pub fn new_or(or: Vec<FilterOperator>) -> Self {
+        FilterOperator::Or(LogicalOr { or })
+    }
+
+    pub fn new_not(not: Vec<FilterOperator>) -> Self {
```

Comment:
I think this should only be a single `FilterOperator` otherwise it's unclear whether the filters are `and` or `or`